src/extension.ts

```diff
@@ -74,13 +98,59 @@ export function activate(context: ExtensionContext): void {
 
 	}));
 
-	// * Find all open d2 files and add to tracker
+	languages.registerDocumentFormattingEditProvider({ language: d2Lang, scheme: "file" }, {
+		provideDocumentFormattingEdits(document: TextDocument): TextEdit[] {
+
+			const editor = window.visibleTextEditors.find(
+				(editor) => editor.document === document
+			);
+
+			if (editor) {
+				documentFormatter.format(editor);
+			}
+
+			return [];
+		}
+	});
+
+	/** Find all open d2 files and add to tracker if they are 
+	 *  open before the extension loads
+	*/
 	workspace.textDocuments.forEach((td: TextDocument) => {
 		if (td.languageId === d2Ext) {
 			previewGenerator.createObjectToTrack(td);
 		}
 	});
+
+	return {
+		// Sets up our ability to render for markdown files
+		extendMarkdownIt(md: any) {
+			return extendMarkdownItWithD2(md);
+		}
+	}
 }
 
 // This method is called when your extension is deactivated
 export function deactivate(): void { }
+
+const pluginKeyword = 'd2';
+
+/**
+ * 
+ * This function will be asked by the Markdown system to render
+ * a d2 snippit in a markdown file 
+ */
+export function extendMarkdownItWithD2(md: any): any {
+	md.use(mdItContainer, pluginKeyword, {});
+
+	const highlight = md.options.highlight;
+	md.options.highlight = (code: string, lang: string) => {
+		if (lang === d2Lang) {
+			return previewGenerator.generateFromText(code);
```

Comment:
I think it might be good to also display the d2 text itself below the image (instead of replacing it?)

![Screen Shot 2023-01-13 at 12 16 29 PM](https://user-images.githubusercontent.com/85081687/212413053-e582c63f-9acc-4279-981f-59fc2036767b.png)

Reply:
We could, but no other render engine does that.  The paradigm is I get called being asked if I know how to render the text within the code block.  If I say 'no', then the markdown engine keeps looking (if it can't find anybody to render it, it just renders the text).  If I answer yes, then I need to return html that looks like the text describes, whatever it might be.

Add this to your markdown file to see another example:

$$P \left( A=2 \, \middle| \, \dfrac{A^2}{B}>4 \right)$$

The LaTex render engine does the same thing.